[AMDGPU] Insert spill codes for the SGPRs used for EXEC copy

cdevadas opened the pull request:
The SGPR registers used for preserving EXEC mask while lowering the whole-wave register spills and copies should be preserved at the prolog and epilog if they are in the CSR range. It isn't happening when there is only wwm-copy lowered and there are no wwm-spills. This patch addresses that problem.

arsenm commented on llvm/lib/Target/AMDGPU/SILowerWWMCopies.cpp:
```diff
@@ -137,5 +137,10 @@ bool SILowerWWMCopies::runOnMachineFunction(MachineFunction &MF) {
+  // SGPRs reserved for exec copy should be preserved as we encountered WWM_COPY
+  // instances.
+  if (Changed)
+    MFI->setPreserveExecCopyReservedReg();
```
Do you really need to track a separate field for this? Can't you just check if RegForExecCopy is used?

cdevadas committed:
[AMDGPU] Insert spill codes for the SGPRs used for EXEC copy

The SGPR registers used for preserving EXEC mask while lowering
the whole-wave register spills and copies should be preserved at
the prolog and epilog if they are in the CSR range. It isn't
happening when there is only wwm-copy lowered and there are no
wwm-spills. This patch addresses that problem.

arsenm commented on llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:
```diff
@@ -1515,11 +1517,15 @@ void SIFrameLowering::determinePrologEpilogSGPRSaves(
                                      /*IncludeScratchCopy=*/false);
     }
+  } else if (ReservedRegForExecCopy) {
+    // Reset it at this point. There are no whole-wave copies and spills
+    // encountered.
+    MFI->setSGPRForEXECCopy(AMDGPU::NoRegister);
```
Does this actually happen? I thought we only reserved this after we know we need it

cdevadas replied:
Yes. It happens when we have the spills inserted during SILowerSGPRSpills and at that point we ensure there is a reserved register for Exec copy. But vgpr-regalloc didn't insert a liverange spill/copy for the wwm-registers.

arsenm reviewed (commented):
Can you add a mir test?

cdevadas committed:
replaceRegWith the shifted register and mir tests.

cdevadas commented:
MIR test added.
